### PR TITLE
TSI Series Block Partitioning

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This changes the `tsi.SeriesBlock` to chunk the hash map index into 64K sections so that the series encoding uses a fixed amount of memory.

###### Required for all non-trivial PRs
- [x] Rebased/mergable
- [x] Tests pass

### CODE REVIEW

File: tsdb/index/tsi1/log_file.go [hunk 1]
```diff
@@ -779,10 +800,8 @@ func (f *LogFile) writeSeriesBlockTo(w io.Writer, info *logFileCompactInfo, n *i
 		mmInfo.seriesIDs = make([]uint64, 0, len(mm.series))
 
 		for _, serie := range mm.series {
-			seriesKey = AppendSeriesKey(seriesKey[:0], serie.name, serie.tags)
-
 			// Lookup series offset.
-			offset := enc.Offset(seriesKey)
+			offset, _ := sblk.Offset(serie.name, serie.tags, seriesKey[:0])
```
Comment (jwilder) [inline tsdb/index/tsi1/log_file.go]: Ah... `_` is `tombstone bool`

File: tsdb/index/tsi1/series_block.go [hunk 2]
```diff
@@ -167,11 +171,28 @@ func (blk *SeriesBlock) UnmarshalBinary(data []byte) error {
...
+
+		// Read block capacity.
+		idx.capacity, buf = int64(binary.BigEndian.Uint64(buf[:8])), buf[8:]
+
+		// Read min key.
+		var n uint64
+		n, buf = binary.BigEndian.Uint64(buf[:8]), buf[8:]
+		idx.min, buf = buf[:n], buf[n:]
+	}
+	assert(len(buf) == 0, "data remaining in index list buffer: %d", len(buf))
```
Comment (jwilder) [inline tsdb/index/tsi1/series_block.go]: Debugging? Remove?

File: tsdb/index/tsi1/series_block.go [hunk 3]
```diff
@@ -473,16 +531,13 @@ func NewSeriesBlockEncoder(w io.Writer) *SeriesBlockEncoder {
...
-func (enc *SeriesBlockEncoder) Offset(key []byte) uint64 {
-	v, _ := enc.offsets.Get(key).(uint64)
-	return v
-}
-
 // Encode writes a series to the underlying writer.
 // The series must be lexicographical sorted after the previous encoded series.
 func (enc *SeriesBlockEncoder) Encode(name []byte, tags models.Tags, deleted bool) error {
+	if len(name) == 0 {
+		panic("empty name")
```
Comment (jwilder) [inline tsdb/index/tsi1/series_block.go]: Return error here.

File: tsdb/index/tsi1/series_block.go [hunk 4]
```diff
@@ -513,6 +573,9 @@ func (enc *SeriesBlockEncoder) Encode(name []byte, tags models.Tags, deleted boo
 	}
 
 	// Save offset to generate index later.
+	if offset <= 0 {
+		panic(fmt.Sprintf("invalid offset: %d", offset))
```
Comment (jwilder) [inline tsdb/index/tsi1/series_block.go]: Return error here.

File: tsdb/index/tsi1/series_block.go [hunk 5]
```diff
@@ -602,6 +679,87 @@ func (enc *SeriesBlockEncoder) ensureHeaderWritten() error {
...
+		if err := writeUint64To(enc.w, uint64(seriesOffset), &enc.n); err != nil {
+			return err
+		}
+	}
+
+	// Determine total size.
+	size := enc.n - offset
+
+	// Verify actual size equals calculated size.
+	assert(size == sz, "series hash index size mismatch: %d <> %d", size, sz)
```
Comment (jwilder) [inline tsdb/index/tsi1/series_block.go]: Debugging?  Remove?

File: tsdb/index/tsi1/index_files.go [hunk 6]
```diff
@@ -227,8 +240,7 @@ func (p IndexFiles) writeTagsetTo(w io.Writer, name []byte, info *indexCompactIn
 			sitr := p.TagValueSeriesIterator(name, ke.Key(), ve.Value())
 			var seriesIDs []uint64
 			for se := sitr.Next(); se != nil; se = sitr.Next() {
-				seriesKey = AppendSeriesKey(seriesKey[:0], se.Name(), se.Tags())
-				seriesID := info.enc.Offset(seriesKey)
+				seriesID, _ := info.sblk.Offset(se.Name(), se.Tags(), seriesKey[:0])
```
Comment (benbjohnson) [inline tsdb/index/tsi1/index_files.go]: This is a `tombstone` flag on the series. It's not needed here.

File: tsdb/index/tsi1/index_files.go [hunk 7]
```diff
@@ -227,8 +240,7 @@ func (p IndexFiles) writeTagsetTo(w io.Writer, name []byte, info *indexCompactIn
...
 				if seriesID == 0 {
 					panic(fmt.Sprintf("expected series id: %s/%s", se.Name(), se.Tags().String()))
```
Comment (benbjohnson) [inline tsdb/index/tsi1/index_files.go]: Converted to an error in 95d4016.

File: tsdb/index/tsi1/series_block.go (hunk 2 above)
Comment (benbjohnson) [inline tsdb/index/tsi1/series_block.go]: Converted to an error in 95d4016.

File: tsdb/index/tsi1/series_block.go (hunk 3 above)
Comment (benbjohnson) [inline tsdb/index/tsi1/series_block.go]: Debugging code. Removed in 95d4016.

File: tsdb/index/tsi1/series_block.go (hunk 4 above)
Comment (benbjohnson) [inline tsdb/index/tsi1/series_block.go]: Debugging code. Removed 95d4016.

File: tsdb/index/tsi1/series_block.go (hunk 5 above)
Comment (benbjohnson) [inline tsdb/index/tsi1/series_block.go]: Converted to an error in 95d4016.

**Pull request merged**